core/account-id/src/lib.rs

```diff
@@ -371,6 +370,36 @@ impl From<AccountId> for Box<str> {
     }
 }
 
+#[cfg(feature = "arbitrary")]
+impl<'a> arbitrary::Arbitrary<'a> for AccountId {
+    fn size_hint(_depth: usize) -> (usize, Option<usize>) {
+        (AccountId::MIN_LEN, Some(AccountId::MAX_LEN))
+    }
+
+    fn arbitrary(u: &mut arbitrary::Unstructured<'a>) -> arbitrary::Result<Self> {
+        let s = u.arbitrary::<&str>()?;
+        match s.parse::<AccountId>() {
+            Ok(account_id) => Ok(account_id),
+            Err(e) => {
+                if let Some((valid_up_to, _)) = e.char {
+                    let valid = &s[..valid_up_to];
+                    debug_assert!(AccountId::validate(valid).is_ok());
```

Comment:
@miraclx I feel this is an invalid assumption. Maybe I miss something, but wouldn’t `a|xyz` fail validation and this code then try to construct account id `a` as valid? (It is too short)

Reply:
This probably should be a fix of `AccountId::validate` (probably coupled with a major release as it’s technically a breaking change, even though it’s actually a bugfix):
- `ParseErrorKind::TooLong` should return a char up-to-which it’s valid
- `ParseErrorKind::{InvalidChar,RedundantSeparator}` should return `None` as a char if it’s actually not possible to make a valid account-id from a substring of the provided string

Does that make sense? (I had assumed that was the way `validate` worked during the review, so if we were two being tricked by this it’s probably worth fixing the root cause, unless there’s a specific reason for the current behavior of `validate`)

Reply:
> `ParseErrorKind::TooLong` should return a char up-to-which it’s valid

I'd argue there's nothing to gain from this information that `AccountId::MAX_LEN` doesn't already provide.

> `ParseErrorKind::{InvalidChar,RedundantSeparator}` should return `None` as a char if it’s actually not possible to make a valid account-id from a substring of the provided string

I'm not convinced this helps either. The point of `validate` is to test if the complete input string slice adheres to all the rules and is safe to be represented as a NEAR `AccountId`. The feedback it provides is to aid eventual modification of the input **by the user**, and not necessarily dynamically by the program.

`validate` returning `InvalidChar` in frol's example only tells you that the second character is invalid. It doesn't provide any guarantees that what has been validated so far (when taken in isolation) will construct a valid ID. Validation is only complete after all the input has been consumed.

To do what you're suggesting, we'll have to do backtracked validation on error detection. And the only utility to that is to slice the input, (like we're doing with this arbitrary implementation). Which is only really useful if the input was computer-generated. `AccountId` inputs are often sourced from humans.

For example, in `a_-b`, `a_` is perfectly valid until we detect that the next character is `-`. We can only return information about the current invalid character, as opposed to running a backtracked validation of the `a_` as well.

We can move this outside though, and do it directly in our arbitrary implementation, in which case, all we have to do is run recursive validation on slices of the input (provided indexes from the validator). Until either the index reaches `0` or the validation returns an `Ok`.

✨: Even a slice of a valid `AccountId` isn't a valid `AccountId` without further validation.

Reply:
> I'd argue there's nothing to gain from this information that AccountId::MAX_LEN doesn't already provide

If eg. the character at MAX_LEN is a `.`, then the char returned would be the one before that.

> I'm not convinced this helps either. The point of validate is to test if the complete input string slice adheres to all the rules and is safe to be represented as a NEAR AccountId. The feedback it provides is to aid eventual modification of the input by the user, and not necessarily dynamically by the program.

I guess the question raised here is what do you want the API of `validate` to actually be. Good news: I just found out that `char` is a private API anyway, so it’s not a big deal, and whichever option we choose here (changing `validate` or changing the `arbitrary` implementation), it’s not going to have any impact in practice.

I guess my takeaway from this discussion would be, we should rename `ParseAccountError::char` into something like `error_on`, so that it’s obvious that it doesn’t indicate the location until which the provided account-id was valid. WDYT?